pkg/application/applyBlueprintSpecUseCase_test.go

```diff
@@ -15,84 +15,126 @@ func TestNewApplyBlueprintSpecUseCase(t *testing.T) {
-		assert.Equal(t, domain.StatusPhaseInProgress, spec.Status)
-		assert.False(t, shouldApply)
+	})
+	t.Run("error activating maintenance mode", func(t *testing.T) {
+		spec := &domain.BlueprintSpec{
+			Status: domain.StatusPhaseEcosystemHealthyUpfront,
+		}
+
+		repoMock := newMockBlueprintSpecRepository(t)
+		maintenanceMock := newMockMaintenanceMode(t)
+		repoMock.EXPECT().GetById(testCtx, blueprintId).Return(spec, nil)
+		maintenanceMock.EXPECT().Activate(mock.Anything).Return(assert.AnError)
+		useCase := NewApplyBlueprintSpecUseCase(repoMock, nil, nil, maintenanceMock)
+
+		err := useCase.PreProcessBlueprintApplication(testCtx, blueprintId)
+
+		require.ErrorIs(t, err, assert.AnError)
+	})
+	t.Run("do nothing on dry run", func(t *testing.T) {
+		spec := &domain.BlueprintSpec{
+			Status: domain.StatusPhaseEcosystemHealthyUpfront,
+			Config: domain.BlueprintConfiguration{DryRun: true},
+		}
+
+		repoMock := newMockBlueprintSpecRepository(t)
+		maintenanceMock := newMockMaintenanceMode(t)
+		repoMock.EXPECT().GetById(testCtx, blueprintId).Return(spec, nil)
+		repoMock.EXPECT().Update(testCtx, spec).Return(nil)
+		useCase := NewApplyBlueprintSpecUseCase(repoMock, nil, nil, maintenanceMock)
+
+		err := useCase.PreProcessBlueprintApplication(testCtx, blueprintId)
+
+		require.NoError(t, err)
+		require.Equal(t, 1, len(spec.Events))
+		assert.Equal(t, domain.BlueprintDryRunEvent{}, spec.Events[0])
 	})
 }
 
-func TestApplyBlueprintSpecUseCase_MarkFailed(t *testing.T) {
+func TestApplyBlueprintSpecUseCase_markInProgress(t *testing.T) {
 	t.Run("ok", func(t *testing.T) {
 		spec := &domain.BlueprintSpec{
-			Status: domain.StatusPhaseInProgress,
+			Status: domain.StatusPhaseEcosystemHealthyUpfront,
 		}
 
 		repoMock := newMockBlueprintSpecRepository(t)
 		repoMock.EXPECT().Update(testCtx, spec).Return(nil)
 		installUseCaseMock := newMockDoguInstallationUseCase(t)
 		useCase := ApplyBlueprintSpecUseCase{repo: repoMock, doguInstallUseCase: installUseCaseMock}
 
-		err := useCase.MarkFailed(testCtx, spec, assert.AnError)
+		err := useCase.startApplying(testCtx, spec)
 
 		require.NoError(t, err)
-		assert.Equal(t, domain.StatusPhaseFailed, spec.Status)
+		assert.Equal(t, domain.StatusPhaseInProgress, spec.Status)
 	})
 
 	t.Run("repo error", func(t *testing.T) {
 		spec := &domain.BlueprintSpec{
-			Status: domain.StatusPhaseInProgress,
+			Status: domain.StatusPhaseEcosystemHealthyUpfront,
 		}
 
 		repoMock := newMockBlueprintSpecRepository(t)
 		repoMock.EXPECT().Update(testCtx, spec).Return(assert.AnError)
 		installUseCaseMock := newMockDoguInstallationUseCase(t)
 		useCase := ApplyBlueprintSpecUseCase{repo: repoMock, doguInstallUseCase: installUseCaseMock}
 
-		err := useCase.MarkFailed(testCtx, spec, assert.AnError)
+		err := useCase.startApplying(testCtx, spec)
 
 		require.ErrorIs(t, err, assert.AnError)
-		assert.Equal(t, domain.StatusPhaseFailed, spec.Status)
+		assert.Equal(t, domain.StatusPhaseInProgress, spec.Status)
```

Comment:
So when this happens in the cluster the blueprintSpec would have the status EcosystemHealthyUpfront, wouldn't it? Because the save of the repo is throwing the error. Would that be problematic? It would just start the blueprint application another time, when it will be reconciled again, wouldn't it? There is just the same problem as with dry-run. If it gets reconciled after a longer pause, all the checks could be invalid in the meantime and will not be checked again.